Feature/zendesk dark mode

khaykov opened the pull request:
Fixes #11111

This PR adds a Dark Mode support to Zendesk activities. Dark Mode is "best-effort" kind of thing - we can do quite a lot of overriding and maybe get 95% the looks we want, but it's a bit "hacky" and quite likely will result in unintended consequences and issues with future upgrades of ZD. I think the look I got is a good compromise.

Apart from a couple of necessary changes, I left the overridden layouts and styles as-is.

I also updated Zendesk to 4.0.0 - only required modification was switching from `UiConfig` to `Configuration`.

![comparison_ps](https://user-images.githubusercontent.com/728822/84431135-a6a35480-abdf-11ea-947a-2fdbb3a0146d.png)


To test:
- Empty tickets list.
- No Connection Snackbar (airplane mode)
- Error when sending a message (airplane mode)
- Sent message.
- Reply from the agent.
- Attachment file.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

khaykov committed:
Updated zendesk version to 4.0.0

khaykov committed:
Overriding main zendesk activities.

khaykov committed:
Updated overriden layouts.

khaykov committed:
Style and color overrides.

khaykov committed:
Added comment about overriden layouts for future upgrades of ZD

khaykov committed:
Updated color.

khaykov committed:
Updated look of request list button.

khaykov committed:
Ignoring some lint warning related to ZD.

khaykov committed:
Merge branch 'develop' of https://github.com/wordpress-mobile/WordPress-Android into feature/zendesk-dark-mode

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/support/ZendeskHelper.kt

khaykov committed:
Ignoring safe InvalidPackage error from Zendesk upgrade.

khaykov committed:
Updated comment, removed newline.

AmandaRiu commented on WordPress/src/main/res/values-night/styles.xml:
```diff
@@ -124,4 +124,58 @@
+        <item name="snackbarStyle">@style/Widget.MaterialComponents.Snackbar</item>
+    </style>
+
+    <style name="ZendeskSdkTheme.Light.RequestList.AvatarText" parent="@style/TextAppearance.MaterialComponents.Body1">
+        <item name="android:textSize">14sp</item>
+        <item name="android:lines">1</item>
+    </style>
+
+    <style name="ZendeskSdkTheme.Light.Request.Message.SystemMessage" tools:ignore="UnusedResources" parent="@style/TextAppearance.MaterialComponents.Body1">
+        <item name="android:lineSpacingExtra">8dp</item>
+        <item name="android:fontFamily">sans-serif</item>
+        <item name="android:textStyle">normal</item>
+    </style>
+
+    <style name="ZendeskSdkTheme.Light.RequestList.Subject" parent="@style/TextAppearance.MaterialComponents.Body1">
+        <item name="android:textSize">14sp</item>
+        <item name="android:lines">1</item>
+    </style>
+
+    <style name="ZendeskSdkTheme.Light.RequestList.Title" parent="@style/TextAppearance.MaterialComponents.Body1">
+        <item name="android:textSize">16sp</item>
+        <item name="android:lines">1</item>
+    </style>
+
+    <style name="ZendeskSdkTheme.Light.Request.Date" parent="@style/TextAppearance.MaterialComponents.Body1">
+        <item name="android:textColor">?attr/wpColorOnSurfaceMedium</item>
+        <item name="android:textSize">14sp</item>
+        <item name="android:fontFamily">sans-serif</item>
+        <item name="android:textStyle">normal</item>
+    </style>
+
+    <style name="ZendeskRequestListButton" parent="@style/WordPress.Button.Primary" />
```
❓ I'm curious why these styles are only overridden in the `values-night` resource file and do not exist in the non-night version. None of these styles are specific to night-mode so if they are moved to the regular `/values/styles.xml` resource file you'd have a more consistent experience between light and dark mode. A good example of this is the differences in the formatting of the system message in your screenshots between light and dark themes:

![Screen Shot 2020-06-26 at 1 45 49 PM](https://user-images.githubusercontent.com/5810477/85895198-6df4a500-b7b3-11ea-91fb-afb8e756e5f6.png)

Moving all the styles that do not specifically override something special for dark mode to the main styles resource file will fix these inconsistencies, for example with the system message referenced above:

Light | dark 
-- | --
![Screenshot_1593200822](https://user-images.githubusercontent.com/5810477/85895270-9c728000-b7b3-11ea-9478-54fccd827544.png)|![Screenshot_1593200827](https://user-images.githubusercontent.com/5810477/85895275-9e3c4380-b7b3-11ea-978c-f7d7dbdd7939.png)

AmandaRiu commented on WordPress/src/main/res/values-night/styles.xml:
```diff
@@ -124,4 +124,58 @@
+        <item name="android:textColor">@color/blue_30</item>
+    </style>
+
+    <style name="ZendeskSdkTheme.Light.Request.Dialog.Retry" parent="ZendeskSdkTheme.Light">
+        <item name="android:clickable">true</item>
+        <item name="android:focusable">true</item>
+        <item name="android:foreground">?selectableItemBackground</item>
+        <item name="android:paddingBottom">12dp</item>
+        <item name="android:paddingTop">12dp</item>
+        <item name="android:paddingEnd">16dp</item>
+        <item name="android:paddingRight">16dp</item>
+        <item name="android:paddingStart">16dp</item>
+        <item name="android:paddingLeft">16dp</item>
+    </style>
+
```
Same as above...this should also be moved to the main styles resource file since it doesn't specify anything specific to dark mode.

AmandaRiu reviewed (changes requested):
@khaykov Finished the first round. It works and actually I think I'll port some of your hacks over to Woo since we've had some additional bugs creep up. It's a tinker toy tower, but not one we have much control over unfortunately. I left a couple code comments and the only other thing I noticed was a lot of hard-coded dimension and font size values. I'm not sure if that's common in this repo so I just wanted to point them out in case they should be replaced with a dimens variable.

khaykov committed:
Merge branch 'develop' of https://github.com/wordpress-mobile/WordPress-Android into feature/zendesk-dark-mode

khaykov committed:
Moved zendesk styles overload from night to normal styles.

khaykov committed:
Cleaned up values.

khaykov commented:
@AmandaRiu Thaks for review and suggestions! You are right, moving styles to regular styles.xml will give it a more consistent look. Looking at my own screenshots it's pretty obvious, no idea what I was thinking back then 🤦 

My original idea was to leave most of the stuff as it was in zendesk, while only modifying what was needed, but. since I already modified parent styles, I went ahead and cleaned up some unnecessary value overloads and extracted a couple of hardcoded ones where it makes sense.

oguzkocer commented:
Hi there! `15.2` is being frozen, so this PR is being bumped to `15.3`. If these changes _have to_ be shipped with `15.2`, please target `release/15.2` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

oguzkocer commented:
@khaykov @AmandaRiu This PR is being bumped a second time without any updates. Could you please either merge it until the next code freeze or update the milestone manually and document the blocker for it? Thank you!

---

Hi there! `15.3` is being frozen, so this PR is being bumped to `15.4`. If these changes _have to_ be shipped with `15.3`, please target `release/15.3` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

_This is a standard release version bump message. Let me know if you have any feedback._

AmandaRiu reviewed (changes requested):
@khaykov changes look good, but during a second round of testing I found that the **Browse our FAQ** doesn't work right in Dark Mode. Not sure how I missed that the first time around 😕 

Dark | Light
-- | --
![Screenshot_1594678716](https://user-images.githubusercontent.com/5810477/87361126-c6a09d80-c528-11ea-83e9-d66bc5a4c9b7.png)|![Screenshot_1594678721](https://user-images.githubusercontent.com/5810477/87361128-c7d1ca80-c528-11ea-9b63-fab42d57b61e.png)

khaykov commented:
Didn't realize that the FAQ screen is also zendesk. I spent some time trying to convert it to Dark Mode but encountered a roadblock. I have an idea how to fix it, but it will take a bit of time.

khaykov commented:
I'm going to close this one for now. Hopefully, I'll be able to get back to is soon.
